### PR TITLE
rule parser - faster parsing by not using pcre - v7

### PR DESCRIPTION
**Pull request by jasonish**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [X] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [X] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [ ] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Previous PRs:
- https://github.com/OISF/suricata/pull/3007
- https://github.com/OISF/suricata/pull/3005

Don't use pcre to parse the rule options at the top level of parsing. Performs much better, especially on lower end CPUs. Could also allow for better context when parsing fails.

Changes from last PR:
- Fix possible out of bounds read (https://github.com/OISF/suricata/pull/3015/files#diff-2a9e3785d8ddf41dcda16958883f970fR575)
- Track length of string to be copied to avoid strdup.

Prscript:
- PR jasonish-pcap: https://buildbot.openinfosecfoundation.org/builders/jasonish-pcap/builds/224
- PR jasonish: https://buildbot.openinfosecfoundation.org/builders/jasonish/builds/577

### CODE REVIEW

File: src/detect-parse.c [hunk 1]
```diff
@@ -973,41 +916,137 @@ static int SigParseAction(Signature *s, const char *action)
...
+        len--;
+    }
+
+    char *endptr = strpbrk(token, " \t\n\r");
+    if (endptr != NULL) {
+        token_len = endptr - token;
+    } else {
+        token_len = len;
+    }
+    strlcpy(output, token, token_len + 1);
```
Comment (victorjulien) [inline src/detect-parse.c]: this is invalid use of strlcpy, you're supposed to pass the output buffer size as length

File: src/detect-parse.c [hunk 2]
```diff
@@ -973,41 +916,137 @@ static int SigParseAction(Signature *s, const char *action)
...
+ * For rule parsing a token is considered to be a string of characters
+ * separated by white space.
+ *
+ * \param input double pointer to input buffer, will be advanced as input is
+ *     parsed.
+ * \param output buffer to copy token into.
+ * \param output_size length of output buffer.
+ */
+static inline int SigParseToken(const char **input, char *output,
+    const size_t output_size)
```
Comment (victorjulien) [inline src/detect-parse.c]: output_size seems unused

File: src/detect-parse.c [hunk 3]
```diff
@@ -973,41 +916,137 @@ static int SigParseAction(Signature *s, const char *action)
...
+            in_list++;
+        } else if (c == ']') {
+            in_list--;
+        } else if (c == ' ') {
+            if (!in_list) {
+                break;
+            }
+        }
+    }
+    strlcpy(output, *input, i + 1);
```
Comment (victorjulien) [inline src/detect-parse.c]: same here

File: src/detect-parse.c [hunk 4]
```diff
@@ -973,41 +916,137 @@ static int SigParseAction(Signature *s, const char *action)
...
+ * Parses rule tokens that may be lists such as addresses and ports
+ * handling the case when they may not be lists.
+ *
+ * \param input double pointer to input buffer, will be advanced as input is
+ *     parsed.
+ * \param output buffer to copy token into.
+ * \param output_size length of output buffer.
+ */
+static inline int SigParseList(const char **input, char *output,
+    const size_t output_size)
```
Comment (victorjulien) [inline src/detect-parse.c]: unused?

File: src/detect-parse.c (hunk 1 above)
Comment (jasonish) [reply src/detect-parse.c]: Perhaps I'll replace the strdup with a copy on the stack, so the NULs can be put back in and strlcpy can be used as intended.

**Comment by jasonish** [comment]
Closing. See https://github.com/OISF/suricata/pull/3020

**Pull request closed without merging**